src/main/java/com/googleintern/wfm/ruleengine/action/service/RuleSuggestion.java

```diff
@@ -0,0 +1,135 @@
+      System.getProperty("user.home")
+          + "/Project/wfm-rule-suggestion-engine/data/support_test_agents_anonymized.csv";
+
+  private static final String CSV_VALIDATION_RESULT_FILE_PATH =
+      System.getProperty("user.home")
+          + "/Project/wfm-rule-suggestion-engine/output/validate_results.csv";
+
+  private static final String CSV_OUTPUT_FILE_PATH =
+      System.getProperty("user.home")
+          + "/Project/wfm-rule-suggestion-engine/output/generated_rules.csv";
+
+  public static void main(String[] args) throws IOException, CsvException {
+    RuleSuggestion ruleSuggestion = new RuleSuggestion();
+    ruleSuggestion.suggestRules(CSV_INPUT_FILE_PATH);
+  }
+
+  /**
+   * suggestRules function is used to generate new rules based on data at the csvFilePath.
+   *
+   * <p>Steps:
+   *
+   * <ol>
+   *   <li>Step 1: Read user data from the input csvFilePath. Save the reading results in list as
+   *       {@link UserModel}.
+   *   <li>Step 2: Filter out valid {@link UserModel} data.
+   *   <li>Step 3: Group valid user data by their work group ID.
+   *   <li>Step 4: Find out general {@link RuleModel} that can cover all {@link UserModel} from the
+   *       same work group ID.
+   *   <li>Step 5: Group existing combinations of {@link FilterModel} from the same work group ID by
+   *       {@link PoolAssignmentModel}.
+   *   <li>Step 6: Run reduction algorithm on each {@link PoolAssignmentModel} and generate {@link
+   *       RuleModel} based on reduced results.
+   *   <li>Step 7: Write generated rules into output file at CSV_OUTPUT_FILE_PATH location.
+   *   <li>Step 8: Valid the performance of generated rules and write validation results at
+   *       CSV_VALIDATION_RESULT_FILE_PATH location.
+   * </ol>
+   */
+  @Override
+  public String suggestRules(String csvFilePath) throws IOException, CsvException {
+    ImmutableList<UserModel> userPoolAssignments = CsvParser.readFromCSVFile(csvFilePath);
+
+    ImmutableList<UserModel> validUserPoolAssignments =
+        DataProcessor.filterValidData(userPoolAssignments);
+
+    ImmutableSet<RuleModel> rules = suggestRules(validUserPoolAssignments);
+
+    CsvWriter.writeDataIntoCsvFile(CSV_OUTPUT_FILE_PATH, rules.asList());
+
+    RuleValidation ruleValidation = new RuleValidation(validUserPoolAssignments);
+    ruleValidation.validate(rules).writeToCsvFile(CSV_VALIDATION_RESULT_FILE_PATH);
+    return null;
+  }
+
+  @Override
+  public String suggestRules(String csvFilePath, int percentage) {
+    return null;
+  }
+
+  private ImmutableSet<RuleModel> suggestRules(ImmutableList<UserModel> validUserPoolAssignments) {
+    ImmutableListMultimap<Long, UserModel> usersByWorkgroupId =
+        WorkgroupIdGroupingUtil.groupByWorkGroupId(validUserPoolAssignments);
+    ImmutableSet.Builder<RuleModel> rulesBuilder = ImmutableSet.builder();
+
+    for (Long workgroupId : usersByWorkgroupId.keySet()) {
+      ImmutableSet<RuleModel> generalRulesForWorkgroupId =
```

Comment:
rename to show that these rules do not have filter and applied to workforce directly.